### PR TITLE
fix: Fix TimePicker filter behavior. Misc ComboBox Fixes

### PR DESCRIPTION
**Pull request by brandonlenz**
# Summary

The MVP TimePicker filtering behavior was identical to our ComboBox implementation, since it was just wrapping the ComboBox.

USWDS made some changes to filtering to the ComboBox (see #1108). This PR also brings those enhancements to our reactified library. 

There was also some existing behavior of our combobox that did not match the USWDS combobox (whether that was always the case, or due to the USWDS changes since our original implementation, I don't know, but I fixed as many as I found here along the way). 

Previous ComboBox behavior can be tested at our hosted storybook: https://trussworks.github.io/react-uswds/?path=/story/components-form-controls-combo-box--default-combo-box-with-prop-options

Current USWDS ComboBox behavior: https://designsystem.digital.gov/components/combo-box/

Previous TimePicker behavior can be tested at our hosted storybook: https://trussworks.github.io/react-uswds/?path=/story/components-form-controls-time-picker--complete-time-picker

Current USWDS TimePicker behavior: https://designsystem.digital.gov/components/time-picker/
 
The following issues were found & resolved (and tests added to hopefully cement the fixes since they're not all obvious):

> **Note:** The videos show the following:
>
> | USWDS Model Behavior | Current Behavior | Fixed Behavior (this PR) |
> | --- | --- | --- |

- Pressing enter on the input without an exact match, but with a previously selected option should revert to the selected option.

https://user-images.githubusercontent.com/15805554/115453292-78be7180-a1ed-11eb-87ac-756e8369caa8.mov

- Typing into the input to filter options, then Blurring without selecting should reset the filter (previously it would keep the filter until the input value was changed)

https://user-images.githubusercontent.com/15805554/115453475-aefbf100-a1ed-11eb-85b8-fe58fbfe7d62.mov

- Selecting an option should give that option the selected style in the dropdown, even if the input filter allows more than one option to be seen.

https://user-images.githubusercontent.com/15805554/115453435-a3102f00-a1ed-11eb-8ef7-649ba5d09686.mov

- When no option is selected, but the dropdown is open, the first option should have the focused style applied (but not actually be focused)

https://user-images.githubusercontent.com/15805554/115453664-dfdc2600-a1ed-11eb-86bf-ef2bb4af3a05.mov

- When the ComboBox is opened and scrolled, then closed, and reopened it should either be at the top, OR if an item is selected, that item should be scrolled into view.

https://user-images.githubusercontent.com/15805554/115453699-e79bca80-a1ed-11eb-949c-16827cd3d027.mov

- When an option is selected, but the filter is changed such that the selected option is not visible, the first partial match should be highlighted with the focused style (but not actually be focused)

https://user-images.githubusercontent.com/15805554/115453740-f5e9e680-a1ed-11eb-8ba6-dd92cb318afe.mov

- TimePicker should not filter via hiding options, but should move the focus styled element to the first match (and scroll to said match)

https://user-images.githubusercontent.com/15805554/115453777-fda98b00-a1ed-11eb-8b82-ab1ecccdf3d6.mov

- TimePicker should pass custom filtering logic into combobox to generate regex to find options. This allows `5:3p` when typed to match `5:30pm`:
image

https://user-images.githubusercontent.com/15805554/115453829-0ef29780-a1ee-11eb-8751-bad4290dc92d.mov

## Related Issues or PRs

closes: #1108

## How To Test

The unit tests I added help to cover the scenarios I fixed.

To compare to our existing storybook or the USWDS implementation, use the deployed storybook or checkout this branch and run storybook locally.

@christopherhuii I would love your help in testing this in a project since I want to make sure I didn't break the blurring behavior and focus trap fixes you made recently (I did some testing locally, but I'm not sure I covered everything)

### CODE REVIEW

File: src/components/forms/ComboBox/utils.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable security/detect-non-literal-regexp */
+/**
+ * Generate a dynamic regular expression based off of a replaceable and possibly filtered value.
+ *
+ * @param {string} filter The base filter to use. May be extended by `extras`.
+ * @param {string} query The value to use in the regular expression
+ * @param {object} extras An object of regular expressions to replace and filter the query
+ */
+export const generateDynamicRegExp = (
```
Comment (brandonlenz) [inline src/components/forms/ComboBox/utils.ts:9]: This is from USWDS

**Comment by brandonlenz** [comment]
Looks like the Happo Diffs are from 
image

**Comment by brandonlenz** [comment]
Added screen recordings to the description to help illustrate the other misc. fixes I made 😅

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 2]
```diff
@@ -873,6 +1018,9 @@ describe('ComboBox component', () => {
 
       expect(getByTestId('combo-box-clear-button')).not.toBeVisible()
       expect(getByTestId('combo-box-option-list').children.length).toEqual(1)
+
+      fireEvent.blur(input)
+      expect
```
Comment (haworku) [inline src/components/forms/ComboBox/ComboBox.test.tsx:1023]: why call `expect` with no value? I've never seen this pattern now I'm curious

File: src/components/forms/ComboBox/ComboBox.tsx [hunk 3]
```diff
@@ -92,20 +103,28 @@ export const ComboBox = ({
...
+  const [state, dispatch] = useComboBox(
+    initialState,
+    options,
+    disableFiltering,
+    filter
+  )
 
   const containerRef = useRef<HTMLDivElement>(null)
-  const itemRef = useRef<HTMLLIElement>(null)
+  const focusedItemRef = useRef<HTMLLIElement>(null)
```
Comment (haworku) [inline src/components/forms/ComboBox/ComboBox.tsx:127]: This renaming is clarifying ⭐

**Review by haworku — APPROVED** [review]
Great work. Whew. The videos helped a lot to understand the expected behavior - really good example of how to make code review clearer.  I'm excited for some of the new behavior added.

Some of the issues with focus I think  _could_ be regressions (specifically the ones about closing/blurring  and then coming back and reopening and certain items being focused or highlighted). That logic gets very nuanced particularly with filtering that's in progress. Either way, its good to have more tests around this and remove some of our 🐛 fix TODOs 😆.

Agree that we should have easi or @christopherhuii test this in their app before merging anything to `main`.

File: src/components/forms/ComboBox/ComboBox.test.tsx (hunk 2 above)
Comment (brandonlenz) [reply src/components/forms/ComboBox/ComboBox.test.tsx:1023]: uuuuuuh 😅 

This looks like I started writing expects for this tests, then saw a squirrel, then came back and committed. Nice catch! I'm surprised it's valid code tbh 😆 

I'll fix this

Comment (haworku) [reply src/components/forms/ComboBox/ComboBox.test.tsx:1023]: yeah it was valid code and I was like -- hmm does this do something I don't know lol

Comment (brandonlenz) [reply src/components/forms/ComboBox/ComboBox.test.tsx]: Fixed in [these two commits](https://github.com/trussworks/react-uswds/pull/1139/files/9ffac9f9a6af90848252c40a167ef87a48b419ca..7812a6dfacfa35877354cefa5b79d3103292ea75)

**Review by haworku — APPROVED** [review]

**Pull request merged**